### PR TITLE
added typing_extensions.Self to _AxesBase.twinx

### LINKED ISSUE #28624: [Bug]: Bad type hint in `_AxesBase.twinx()`
### Bug summary

Incorrect type annotation for `_AxesBase.twinx()` (should be `Self`, not `_AxesBase`).

### Code for reproduction

```Python
import matplotlib.pyplot as plt
import numpy as np

# First create some toy data:
x = np.linspace(0, 2 * np.pi, 400)
y = np.sin(x**2)
z = np.cos(x**2)

# Create four polar Axes and access them through the returned array
fig, axsA = plt.subplots()
axsA.plot(x, y)
axsB = axsA.twinx()
axsB.plot(x, z)  # <-- Cannot access attribute "plot" for class "_AxesBase"
```

### Actual outcome

axsB is of type `_AxesBase`

### Expected outcome

The type of `axsB` should be the same as the type of `axsA`.

### Matplotlib Version

3.9.1

### PR DESCRIPTION
**Pull request by randolf-scholz**
## PR summary
- adds `typing_extensions >= 4.0.0` as a dependency
- annotates `_AxesBase.twinx` and `_AxesBase.twiny` with return type `typing_extensions.Self`
- [x] closes #28624 
- [N/A] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/api_changes.html#announce-changes-deprecations-and-new-features)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -28,6 +28,7 @@ dependencies:
   - python-dateutil>=2.1
   - setuptools_scm
   - wxpython
+  - typing-extensions>=4.0.0
```
Comment (QuLogic) [inline environment.yml:31]: We already use `typing-extensions` for `Self`; this addition is not needed as a runtime requirement. It is pulled in by mypy.

File: pyproject.toml [hunk 2]
```diff
@@ -39,6 +39,7 @@ dependencies = [
     "pillow >= 8",
     "pyparsing >= 2.3.1",
     "python-dateutil >= 2.7",
+    "typing-extensions >= 4.0.0",
```
Comment (QuLogic) [inline pyproject.toml:42]: Also not needed.

File: environment.yml (hunk 1 above)
Comment (randolf-scholz) [reply environment.yml:31]: There are other type checkers than `mypy`, for instance `pyright` does not have any python dependencies as it is written in TS. Type-checkers may also run in external virtual environments, such as pre-commit hooks.

IMO if someone created a bare venv and only pip installed matplotlib, then stuff should still type check.

Comment (randolf-scholz) [reply environment.yml:31]: I will double-check if it is needed tomorrow.

Comment (QuLogic) [reply environment.yml:31]: This file is not for venvs, but conda; we don't have any typing-related dependencies here. They're handled by pre-commit or the tox environment.

**Comment by ksunden** [comment]
Additionally, we are now python 3.10+ on main anyway, so not extensions is not actually needed for Self.

Note however that to be backported to the next micro release, we would need to take care there, since it still runs on py 3.9.

**Comment by randolf-scholz** [comment]
> Additionally, we are now python 3.10+ on main anyway, so not extensions is not actually needed for Self

`typing.Self` only became part of the standard library in 3.11

File: environment.yml (hunk 1 above)
Comment (randolf-scholz) [reply environment.yml:31]: OK, so I tested it: downloaded current release from PyPI and installed in a bare virtualenv together with pyright (python wrapper)

```
$ pip list
Package         Version
--------------- -----------
contourpy       1.2.1
cycler          0.12.1
fonttools       4.53.1
kiwisolver      1.4.5
matplotlib      3.9.1
nodeenv         1.9.1
numpy           2.0.1
packaging       24.1
pillow          10.4.0
pip             24.2
pyparsing       3.1.2
pyright         1.1.373
python-dateutil 2.9.0.post0
setuptools      68.2.2
six             1.16.0
wheel           0.41.2
```

Seems like it works after doing the minimal edit, despite `typing-extensions` not being present in the virtual environment.

**Comment by ksunden** [comment]
Ah, I remembered Self coming in (at least added in some places) with recent changes (#28503) that were broadly related to py 3.10 being the new minimum.

Regardless, relying on typing_extensions in `pyi` files (or guarded by `if TYPE_CHECKING`) is not new here, though perhaps we could be more explicit about it somewhere.

I do not think we want to make it a _runtime_ dependency necessarily.

All that said, in this instance, I don't actually think `Self` describes what actually happens...

For twinx/twiny, the Axes that is returned is created by `Figure.add_subplot` or `Figure.add_axes`, which both return `Axes`, specifically.

This means that if you call `twinx` on a Polar plot (or geographic projection or 3D plot, etc) you get a "normal" axes back... (See #12506).

Thus the actual correct type hint is `Axes` here, not `Self`:

Created here:

https://github.com/matplotlib/matplotlib/blob/44be14cc3a866e28a9ad22f36abfe8c0623cc6ac/lib/matplotlib/axes/_base.py#L4522-L4525

Type hints for those methods:

https://github.com/matplotlib/matplotlib/blob/44be14cc3a866e28a9ad22f36abfe8c0623cc6ac/lib/matplotlib/figure.pyi#L70-L93

This generally falls into the category of `_AxesBase` is not _totally_ wrong... `Axes` is a subclass thereof, but `Self` is also not correct when you go further down the inheritance tree.


However... I think `Self` may be the "ideal" type hint, just not actually reflective of the code as it stands...

**Comment by timhoffm** [comment]
> This means that if you call `twinx` on a Polar plot (or geographic projection or 3D plot, etc)

It's generally not clear what twinning should mean on arbitrary axes or that it's useful. For example, where should the second radial axis on polar be placed? Polar does not even have a second radial axis. While it would be technically feasible for 3d and geo axis, the resulting plot is likely very difficult to interpret.

For now, I'd pragmatically say that we just don't provide proper twinning for these niche cases and set the return type to `Axes`.

**Comment by QuLogic** [comment]
Thanks @randolf-scholz! Congratulations on your first PR to Matplotlib :tada: We hope to hear from you again.

**Pull request merged**